Eliminate deprecated lib records apis

hnakamur opened the pull request:
This pull request is meant to resolve [Eliminate deprecated lib records APIs · Issue #4468 · apache/trafficserver](https://github.com/apache/trafficserver/issues/4468)

bryancall commented:
[approve ci autest 2]

bryancall commented:
Do we want to return value instead of passing in a pointer and return a std::string_view instead of a char*?

hnakamur commented:
## `RecEstablishStaticConfig*` family needs a pointer to the config value

`RecEstablishStaticConfig*` family have to take a pointer since `RecLinkConfig*` family take a pointer.
https://github.com/apache/trafficserver/blob/91748d47712482fe39dbdfc5c29d46fc65f7d3ae/include/records/RecCore.h#L116-L122

For example, `RecEstablishStaticConfigStringAlloc` calls `RecLinkConfigString`.
https://github.com/hnakamur/trafficserver/blob/4d8bd3638cc0683b12773bedf0a6a27df6e109af/src/records/RecCore.cc#L915-L922
```
RecErrT
RecEstablishStaticConfigStringAlloc(const char *name, RecString *rec_string, bool lock)
{
  if (RecLinkConfigString(name, rec_string) == REC_ERR_OKAY) {
    ats_free(*rec_string);
  }
  return RecGetRecordStringOrNullptr_Xmalloc(name, rec_string, lock);
}
```

## other functions can be changed to return the config value and take a pointer to `RecErrT err` or `bool found`

As for other functions, there are few use cases which needs to know wether the config is found or not, for example:
https://github.com/hnakamur/trafficserver/blob/4d8bd3638cc0683b12773bedf0a6a27df6e109af/src/mgmt/config/FileManager.cc#L280-L281
```
int  enabled;
bool found = RecGetRecordIntOrZero("proxy.config.body_factory.enable_customizations", &enabled) == REC_ERR_OKAY;
```

We can change RecGetRecordIntOrZero to return the config value and take a pointer to RecErrT instead.
```
RecInt RecGetRecordIntOrZero3(const char *name, RecErrT *err = nullptr, bool lock = true);
```

```
RecInt
RecGetRecordIntOrZero2(const char *name, RecErrT *err, bool lock)
{
  RecData data;
  RecErrT rec_err = RecGetRecord_Xmalloc(name, RECD_INT, &data, lock);
  if (err) {
    *err = rec_err;
  }
  return rec_err == REC_ERR_OKAY ? data.rec_int : 0;
}
```
However we need one more line after rewriting to use this version.
```
RecErrT rec_err;
int     enabled = RecGetRecordIntOrZero2("proxy.config.body_factory.enable_customizations", &rec_err);
bool    found   = rec_err == REC_ERR_OKAY;
```

Maybe we can take a pointer to bool instead of RecErrT.

```
RecInt RecGetRecordIntOrZero3(const char *name, bool *found = nullptr, bool lock = true);
```
```
RecInt
RecGetRecordIntOrZero3(const char *name, bool *found, bool lock)
{
  RecData data;
  RecErrT rec_err = RecGetRecord_Xmalloc(name, RECD_INT, &data, lock);
  if (found) {
    *found = rec_err == REC_ERR_OKAY;
  }
  return rec_err == REC_ERR_OKAY ? data.rec_int : 0;
}
```

```
bool found;
int  enabled = RecGetRecordIntOrZero2("proxy.config.body_factory.enable_customizations", &found);
```

I welcome other suggestions if any.

hnakamur commented:
Another candidate would be returning `std::pair`.
For example, return `std::pair<RecFloat, RecErrT>` from `RecGetRecordFloat` like https://github.com/hnakamur/trafficserver/commit/132006109f3c1ca0cb6508fca81fdf8339a22323

hnakamur commented:
Also, how about change the type of the `name` argument from `const char *` to `const std::string_view &`?

hnakamur commented:
I changed the signature of `RecGetRecordString_Xmalloc` to
```
td::pair<std::string_view, RecErrT> RecGetRecordString_Xmalloc(const char *name, bool lock = true);
```
at https://github.com/hnakamur/trafficserver/compare/94a72272fcfe291ad5710b44b1c3069c94c27581...eliminate_deprecated_lib_records_apis_std_string_view
However I don't think it's good since we need a lot of `const_cast<char *>` to free the allocated string.

So I thought it would be better to return `std::string`.
However `std::string{nullptr, 0}` will be prohibited in C++23 (https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2166r1.html)
Therefore we have to use `std::optional<string>` to distinguish `nullptr` and an empty string.

So I changed the signature of the above function to
```
std::pair<std::optional<std::string>, RecErrT> RecGetRecordStringAlloc(const char *name, bool lock = true);
```
and also rename it to RecGetRecordStringAlloc since it does not use malloc anymore.
This version is at https://github.com/hnakamur/trafficserver/compare/94a72272fcfe291ad5710b44b1c3069c94c27581...eliminate_deprecated_lib_records_apis_std_optional_string

I changed the signature for integer and float version to return std::pair too.
```
std::pair<RecInt, RecErrT>                     RecGetRecordInt(const char *name, bool lock = true);
std::pair<RecFloat, RecErrT>                   RecGetRecordFloat(const char *name, bool lock = true);
RecErrT                                        RecGetRecordString(const char *name, char *buf, int buf_len, bool lock = true);
std::pair<std::optional<std::string>, RecErrT> RecGetRecordStringAlloc(const char *name, bool lock = true);
std::pair<RecCounter, RecErrT>                 RecGetRecordCounter(const char *name, bool lock = true);
```

I also changed `RecEstablishStaticConfigXxx` to inline functions like below:
```
inline RecErrT RecEstablishStaticConfigInt(RecInt &rec_int, const char *name, bool lock = true);
inline RecErrT RecEstablishStaticConfigInt32(int32_t &rec_int, const char *name, bool lock = true);
inline RecErrT RecEstablishStaticConfigUInt32(uint32_t &rec_int, const char *name, bool lock = true);
inline RecErrT RecEstablishStaticConfigString(RecString &rec_string, const char *name, bool lock = true);
inline RecErrT RecEstablishStaticConfigFloat(RecFloat &rec_float, const char *name, bool lock = true);
inline RecErrT RecEstablishStaticConfigByte(RecByte &rec_byte, const char *name, bool lock = true);
```
I also added the following utility functions:
```
inline char *ats_stringdup(std::optional<std::string> const &p);
inline const char *ats_as_c_str(std::optional<std::string> const &p);
```

Could you give me feedback about above changes?
I'll make a proposal on ats-dev mailing list after waiting for a few days.

hnakamur commented:
If we can change the default value of [`proxy.config.diags.action.tags`](https://github.com/apache/trafficserver/blob/e4fab322b45c0250e0a019b41cb590ef5ee521df/src/records/RecordsConfig.cc#L205) from `nullptr` to an empty string `""`, I think we want more simpler APIs like:

```
std::optional<RecInt>      RecGetRecordInt(const char *name, bool lock = true);
std::optional<RecFloat>    RecGetRecordFloat(const char *name, bool lock = true);
RecErrT                    RecGetRecordString(const char *name, char *buf, int buf_len, bool lock = true);
std::optional<std::string> RecGetRecordStringAlloc(const char *name, bool lock = true);
std::optional<RecCounter>  RecGetRecordCounter(const char *name, bool lock = true);
```
`std::optional::has_value()` for the result returns true if the value is found or false otherwise.

This version is at https://github.com/hnakamur/trafficserver/compare/94a72272fcfe291ad5710b44b1c3069c94c27581...eliminate_deprecated_lib_records_apis_std_optional_string_without_error

I found that I forgot to modify unit tests. I'll fix those tommorrow.

hnakamur commented:
I realized that it's OK because the tests I thought I forgot to modify were deleted in [Cleanup - remove a few unused files. by brbzull0 · Pull Request #12061 · apache/trafficserver](https://github.com/apache/trafficserver/pull/12061).

cmcfarlen commented:
I prefer the optional returns so we can use `value_or` now and `transform`, `or_else` in the future.  I wonder, could the string version return an optional string_view of the buffer passed in?

hnakamur commented:
Thank you for a great idea!

I changed
```
RecErrT                    RecGetRecordString(const char *name, char *buf, int buf_len, bool lock = true);
```
to
```
std::optional<std::string_view> RecGetRecordString(const char *name, char *buf, int buf_len, bool lock = true);
```
and updated my eliminate_deprecated_lib_records_apis_std_optional_string_without_error branch.

The diff is https://github.com/hnakamur/trafficserver/compare/94a72272fcfe291ad5710b44b1c3069c94c27581...eliminate_deprecated_lib_records_apis_std_optional_string_without_error

cmcfarlen commented:
We talked about this in the review meeting and we agreed that we like the API that returns an optional.  Could you migrate that PR to be in line with your alternate one?  Thank you for working on this!

hnakamur commented:
Thank you! I've updated this branch to be aligned with the following APIs:
```
std::optional<RecInt>           RecGetRecordInt(const char *name, bool lock = true);
std::optional<RecFloat>         RecGetRecordFloat(const char *name, bool lock = true);
std::optional<std::string_view> RecGetRecordString(const char *name, char *buf, int buf_len, bool lock = true);
std::optional<std::string>      RecGetRecordStringAlloc(const char *name, bool lock = true);
std::optional<RecCounter>       RecGetRecordCounter(const char *name, bool lock = true);
```

bryancall commented:
[approve ci]

bryancall commented:
[approve ci]

bryancall commented:
@hnakamur Can you access this link? https://ci.trafficserver.apache.org/job/Github_Builds/job/clang-analyzer/5809//artifact/output/12060/scan-build-2025-05-05-22-24-53-908747-93rhdmg_/report-0198f5.html#EndPath

It has the Clang Analyzer report

hnakamur commented:
> @hnakamur Can you access this link? https://ci.trafficserver.apache.org/job/Github_Builds/job/clang-analyzer/5809//artifact/output/12060/scan-build-2025-05-05-22-24-53-908747-93rhdmg_/report-0198f5.html#EndPath
> 
> It has the Clang Analyzer report

@bryancall Thank you. I fixed it in [6928f15](https://github.com/apache/trafficserver/pull/12060/commits/6928f15d3a661b3e19d60ede4f796bf7895b6093).
